### PR TITLE
Treat unencoded `/` inside quoted literals as data (not separators)

### PR DESCRIPTION
**Pull request by gathogojr**
## Issues

*This pull request fixes https://github.com/OData/AspNetCoreOData/issues/709, fixes https://github.com/OData/odata.net/issues/2292, fixes https://github.com/OData/WebApi/issues/2309, fixes https://github.com/OData/WebApi/issues/1964, fixes https://github.com/OData/odata.net/issues/2477, fixes https://github.com/OData/odata.net/issues/735, fixes https://github.com/OData/odata.net/issues/1708, fixes https://github.com/OData/odata.net/issues/3327

## Description

This change updates `ParsePathIntoSegments` to:

- Parse the escaped path and split only on unquoted `/`, preserving any unencoded slashes inside quoted literals as part of the same segment for compatibility with some external clients - **reported case referred to Excel as one such integration https://github.com/OData/AspNetCoreOData/issues/709**.  
- Keep the existing behavior for properly **percent‑encoded** slashes (`%2F`) and for balanced quotes (including doubled single quotes for escaping).  
- Preserve prior error behavior for malformed percent escapes and unbalanced quotes.  

This aligns segment parsing with the [OData URL Conventions](http://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part2-url-conventions/odata-v4.0-errata03-os-part2-url-conventions-complete.html#_Toc453752335) (slashes that are part of a literal/key **must** be percent‑encoded) and avoids producing OData‑inapplicable segment shapes when a client violates that rule. 

### Motivation & Standards Alignment

- **OData 4.01 URL Conventions** require that forward slashes in key value segments be percent‑encoded. If a client sends `/People('foo/bar')`, the slash inside the literal should have been `%2F`. When this is not respected, treating the raw `/` as a separator results in segments that **do not correspond to any valid OData path tokenization**. This PR handles that non‑conformance more gracefully by preserving the literal as a single segment.   
- **RFC 3986** defines `/` as a path segment delimiter. Parsing should use the escaped path so that `%2F` stays data while raw `/` remains a separator. This change does exactly that: we operate on the escaped path and only treat unquoted `/` as a separator. 

#### Before (current behavior with non‑conformant clients)
- Request:  
  `GET /Service.svc/People('foo/bar')` *(slash not encoded within the literal)*
- Parsed segments:  
  `["People('foo", "bar')"]` <-- Two segments that are not meaningful OData segments.

#### After (this PR)
- Same request:  
  `GET /Service.svc/People('foo/bar')`
- Parsed segments:  
  `["People('foo/bar')"]`  <-- Single segment, quote‑consistent and usable for OData path handling.

#### No change for conformant inputs
- Request:  
  `GET /Service.svc/People('foo%2Fbar')`
- Parsed segments (before and after):  
  `["People('foo%2Fbar')"]`  <-- slash remains within the literal; downstream can unescape as needed.  
  *(OData 4.01 explicitly requires this encoding.)* 

### Is this a breaking change?

**Technically, yes**  - code that intentionally depended on the previously split result (`["People('foo", "bar')"]`) would now see a single segment. However:

- That two‑segment result was an inapplicable outcome for OData: neither segment is a valid OData path segment on its own; routing, model binding, and `ODataUriParser` semantics do not interpret `"People('foo"` or `"bar')"` as meaningful tokens. In other words, the prior behavior produced an invalid tokenization of an already non‑conformant URL. 
- The new behavior does not loosen standards: it continues to treat raw `/` as a separator, but only when outside a quoted literal. It simply avoids making the situation *worse* when a client neglected to encode the slash that is clearly intended as data. This better matches the intent of RFC 3986’s delimiter rules and keeps `%2F` vs `/` semantics intact. 

**Conclusion:** While “breaking” in a narrow sense, the previous result was not a useful contract to depend on. The new behavior is a corrective change that improves interoperability without changing the meaning of any valid OData URL.

### Additional arguments in favor

1. **Spec‑conformant by default; robust by necessity**  
   We continue to encourage clients to percent‑encode slashes inside literals (as required by OData 4.01). But servers in the wild must interoperate with clients that don’t always follow this strictly. Preserving the literal as a single segment is the most faithful interpretation of the client’s intent and avoids nonsensical segmentation. 

2. **No ambiguity introduced**  
   - We only treat `/` as a separator outside quotes.  
   - Within quotes, we preserve the raw text (including any `/`) and still honor `%27` / `%27%27` and doubled `'` escaping semantics.  
   - Balance checks ensure we error on unbalanced quotes or malformed `%` sequences as before.  
   This makes the change predictable and bounded to a well‑defined compatibility corner case. *(RFC 3986 supports component‑aware parsing; we apply it at the path level.)* 

3. **No change for encoded data**  
   Any client already sending `%2F` inside the literal is unaffected (same result before and after). 

### Risk analysis

**Who might be affected?**
- Downstream code that *explicitly relied* on the old, split behavior (e.g., custom middleware that expects two segments when encountering a raw `/` inside quotes) could observe different segment counts.  
- Security posture: By treating the entire quoted token as one segment, we’re not widening traversal or normalization issues; we still operate on the escaped path and only accept `/` as a separator outside quotes. This reduces ambiguous parses compared to the old behavior that split mid‑literal. 

**Likelihood & impact**
- Likelihood: low - the previous two‑segment shape was not a valid OData path tokenization, so relying on it was unlikely and fragile.  
- Impact: low‑to‑moderate - if such reliance exists, migration is straightforward (treat the quoted literal as a single segment and apply custom parsing inside the literal if truly needed).

Depending on how the maintainers view this change, we can:
- Ship it in a minor release - Would it be worth putting it behind a feature flag?, or  
- Defer the change to the **next major release - ODL 9** with a deprecation note now.  
- My personal recommendation is to merge as a **bug fix** that corrects an inapplicable prior outcome and improves standards alignment with OData and RFC 3986. 

### Memory & performance benchmarks

| Method         | Path                                  | Mean        | Error     | StdDev     | Median      | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|--------------- |-------------------------------------- |------------:|----------:|-----------:|------------:|------:|--------:|-------:|----------:|------------:|
| Parse_Original | Customers                             | 1,508.55 ns | 37.259 ns | 106.902 ns | 1,462.45 ns |  1.00 |    0.10 | 0.1564 |    1984 B |        1.00 |
| Parse_Revised  | Customers                             |    73.09 ns |  1.766 ns |   5.039 ns |    70.40 ns |  0.05 |    0.00 | 0.0254 |     320 B |        0.16 |
|                |                                       |             |           |            |             |       |         |        |           |             |
| Parse_Original | Customers('abc')                      | 1,407.93 ns | 27.255 ns |  27.989 ns | 1,397.31 ns |  1.00 |    0.03 | 0.1640 |    2080 B |        1.00 |
| Parse_Revised  | Customers('abc')                      |    85.10 ns |  3.020 ns |   8.666 ns |    81.30 ns |  0.06 |    0.01 | 0.0286 |     360 B |        0.17 |
|                |                                       |             |           |            |             |       |         |        |           |             |
| Parse_Original | Customers('abc/pqr')                  | 1,684.34 ns | 20.034 ns |  15.641 ns | 1,687.56 ns |  1.00 |    0.01 | 0.1717 |    2168 B |        1.00 |
| Parse_Revised  | Customers('abc/pqr')                  |    99.18 ns |  3.027 ns |   8.437 ns |    95.80 ns |  0.06 |    0.01 | 0.0305 |     384 B |        0.18 |
|                |                                       |             |           |            |             |       |         |        |           |             |
| Parse_Original | Customers('abc%2Fpqr')                | 1,634.20 ns | 32.633 ns |  90.967 ns | 1,593.80 ns |  1.00 |    0.08 | 0.1736 |    2200 B |        1.00 |
| Parse_Revised  | Customers('abc%2Fpqr')                |   119.83 ns |  2.539 ns |   7.202 ns |   117.24 ns |  0.07 |    0.01 | 0.0370 |     464 B |        0.21 |
|                |                                       |             |           |            |             |       |         |        |           |             |
| Parse_Original | Customers('O''Neil')/Orders(42)/Items | 1,595.63 ns | 31.785 ns |  82.047 ns | 1,556.91 ns |  1.00 |    0.07 | 0.1907 |    2400 B |        1.00 |
| Parse_Revised  | Customers('O''Neil')/Orders(42)/Items |   153.86 ns |  3.106 ns |   8.128 ns |   151.02 ns |  0.10 |    0.01 | 0.0427 |     536 B |        0.22 |
|                |                                       |             |           |            |             |       |         |        |           |             |
| Parse_Original | Customers(1)                          | 1,441.38 ns | 35.204 ns |  99.293 ns | 1,419.87 ns |  1.00 |    0.09 | 0.1602 |    2032 B |        1.00 |
| Parse_Revised  | Customers(1)                          |    83.49 ns |  3.155 ns |   9.052 ns |    81.80 ns |  0.06 |    0.01 | 0.0267 |     336 B |        0.17 |
|                |                                       |             |           |            |             |       |         |        |           |             |
| Parse_Original | People(%27O%27%27Neil%27)/Orders(1)   | 1,824.30 ns | 36.504 ns |  94.228 ns | 1,783.86 ns |  1.00 |    0.07 | 0.1888 |    2392 B |        1.00 |
| Parse_Revised  | People(%27O%27%27Neil%27)/Orders(1)   |   166.14 ns |  3.243 ns |   7.451 ns |   162.98 ns |  0.09 |    0.01 | 0.0432 |     544 B |        0.23 |

### Tests
- Adequate tests were added to ensure maximum coverage

## Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### CODE REVIEW

**Comment by xuzhg** [comment]
@gathogojr 
How about the following scenarios:

### /customers('abc/efg') 
  ==>Basic scenario to support, It should work.
### /customers('abc%2F/''efg')/  abc   /'%2F
  ==> Mixed the %2F and / together?
### /customers(%27abc%2F/'%27efg')/  abc'   /'%2F 
 ==> Be noted, how to distinguish the / between abc' and the last '%2f

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs [hunk 1]
```diff
@@ -38,66 +39,198 @@ public UriPathParser(ODataUriParserSettings settings)
         /// <returns>List of unescaped segments.</returns>
         public virtual ICollection<string> ParsePathIntoSegments(Uri fullUri, Uri serviceBaseUri)
```
Comment (habbes) [inline src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:41]: If we want to also take this opportunity to reduce overhead of this method, I'd propose that instead of returning a `ICollection<string>`, which will require us to prematurely split strings, we return a customized collection, e.g. `UriPathSegments` collection which encapsulates the position and length of each segment, this assumes each segment is a substring of the original uri path. This assumption doesn't always hold cause unescaping the string might produce a different string that isn't a substring of the full path. If this path is less common than the path that does not need escaping, then we might have 2 different code paths.

A possible representation is for the collection to be a thin struct wrapper over a the full string path and the a collection of ranges
```c#
struct UriPathSegmentCollection
{
   private string path;
   private List<Range> segmentRanges; // could also be Range[] array if we can get the exact number of segments
   ReadOnlySpan<char> this[int index] {} // .. index based access
   Enumerator GetEnumerator();

   struct Enumerator // supports allocation-free foreach
   {
   }
}
```

Ranges allow you to construct a `ReadOnlySpan<char>` from `path` for each segment on demand. This assumes every segment is a substring of path.

Another representation is to replace the collection of `Range` with a collection of `ReadOnlyMemory`. The `Range` consists of 2 ints wheras a `ReadOnlyMemory` consists of an object reference and 2 ints so it's twice the size of `Range`. But with `ReadOnlyMemory` we would not need to store the `path` property since each segment would have its own reference, which could either be the original path or a custom string created from the escaping operation.

```c#
struct UriPathSegmentCollection
{
   private List<ReadOnlyMemory> segments; 
   // other members: indexer, enumerator, length, etc.
}
```

The latter is more flexible but takes up more space. Benchmarks needed for different operations to evaluate which one is generally better for the use cases we have.

If we find that the second representation has noticeable overhead relative to the first, we could support both, turn them into classes and expose a common interface. Then the parser would select the appropriate representation based on whether the path includes characters than needs escaping. But we would also have to weigh this against the overhead of using a class and the indirection from the interface. If it's not worth it, we'd settle on whether the `ReadOnlyMemory` approach is an acceptable compromise.

Comment (gathogojr) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:41]: @habbes Thanks for the suggestion about deferring materialization. For this change I am intentionally keeping `ParsePathIntoSegments` returning `ICollection<string>` for these reasons:

1. **API contract / compatibility**
This method is public and existing downstream code expects fully materialized, decoded strings. Changing the return type now would be a breaking behavioral & shape change.

2. **Semantics require early normalization**
The parser must decide (while scanning) whether a slash is a segment delimiter. That decision depends on balanced parentheses and correctly terminated single-quoted literals after percent-decoding. Deferring decoding would push that logic onto every consumer or force re‑parsing, negating much of the proposed efficiency.

3. **Escaping prevents a pure slice model**
Segments that contain percent-escapes (or mixed raw/encoded quote forms) are not always substrings of the original path once normalized. Unescaping can:
    - Shorten length (e.g., `%2F`, `%27`, multibyte UTF‑8 sequences).
    - Merge multi-escape sequences into a single code point. Therefore a “range into original string” representation is only valid on the no-escape path. The mixed model (sometimes slice, sometimes new string) adds branching and complexity for a relatively small memory win (typical OData paths are short).
 
4. **Allocation profile is already small after the change - benchmarks further below**
A typical request would yield a few segments; most segments would have no % and would be short. The dominant cost in end-to-end request handling (metadata lookup, binding, materialization) dwarfs a handful of string allocations. Prematurely optimizing here adds maintenance cost and parsing risk in a security-sensitive area (URL handling) for marginal gain.

5. **Correctness > micro-optimization**
Recent changes tightened correctness around malformed / mixed encodings and unterminated quotes. Introducing a lazy abstraction right now would expand the test surface area while we’re still stabilizing corner cases.

That said, I agree there is future potential for a non-allocating path abstraction (especially if we later rework higher layers to operate on spans). I propose we land the correctness-focused change set as-is, then (if profiling shows benefit) add an internal fast-path API that returns a lightweight segment view struct, with the current public API layered on top.

Comment (habbes) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:41]: @gathogojr 

I agree with your conclusion, but thought I'd still reply to some of the comments you made.

**Api contract/compatibility**: my assumption was that this was being done in a breaking change release (ODL 9), I didn't realize it was targeting ODL 8. I also reasoned that this is not an API that's widely used by library users other than our own libraries, so the impact of breaking the contract would not be significant to library consumers.

**Semantics and normalization**: This would not change. I did not propose that we blindly split by the slash. I'd expect we still scan the url character and determine where the boundaries are, this would take quotes and a parantheses into account.

**Escaping prevents a pure slice model**: The proposal also account for it. If a string has not escape sequences, we would slice into the source string, if the string does have an escape sequences, the segments that need to be escaped would allocate a new (escaped) string. That's why I propose `ReadOnlyMemory` which provides flexibility to support both scenarios.

**Allocation profile**: I'd assume that this depends on the length of the url and the number of segments in the URL as well as the rate of incoming URLs we'd need to process. Depending on the scenario, an additional boost may or may not be worthwhile. That said your changes do result in a huge increase from the existing implementation.

**Correctness** I think the same effort for correctness taken in this PR would apply to the proposed changes. You could implement the same logic as you're doing in this PR, what changes is the data structure.

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs [hunk 2]
```diff
@@ -38,66 +40,357 @@ public UriPathParser(ODataUriParserSettings settings)
...
-                serviceBaseUri = UriUtils.CreateMockAbsoluteUri();
-                fullUri = UriUtils.CreateMockAbsoluteUri(fullUri);
+                serviceBaseUri = UriUtils.CreateMockAbsoluteUri(); // -> "http://host/"
+                fullUri = UriUtils.CreateMockAbsoluteUri(fullUri); // -> absolute
+            }
+            else
+            {
+                // Ensure absolute URIs before base-of validation; otherwise
+                // UriUtils.UriInvariantInsensitiveIsBaseOf can throw when given relative URIs.
+                if (!serviceBaseUri.IsAbsoluteUri)
```
Comment (habbes) [inline src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:56]: Does this change existing behaviour?

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs [hunk 3]
```diff
@@ -38,66 +40,357 @@ public UriPathParser(ODataUriParserSettings settings)
...
+                    {
+                        exclusiveEnd = j;
+                        return true; // Found the matching closing paren
+                    }
+
+                    // Unbalanced or unterminated quote -> fail (caller will treat chars normally)
+                    return false;
+                }
+
+                void AddSegmentIfAny()
```
Comment (habbes) [inline src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:304]: Why not make this `static` too and pass the necessary parameters to it? I've seen that this pattern would allocate a struct to store the state, so maybe it's cheap, but I wonder if that's more overhead than just passing the parameters directly.

```c#
class Foo
{
    void Do()
    {
        
        StringBuilder sb = new();
        Help();
        
        void Help()
        {
            sb.Append("what");
        }
    }
}
```
Is lowered to:

```c#
internal class Foo
{
    [StructLayout(LayoutKind.Auto)]
    [CompilerGenerated]
    private struct <>c__DisplayClass0_0
    {
        public StringBuilder sb;
    }

    private void Do()
    {
        <>c__DisplayClass0_0 <>c__DisplayClass0_ = default(<>c__DisplayClass0_0);
        <>c__DisplayClass0_.sb = new StringBuilder();
        <Do>g__Help|0_0(ref <>c__DisplayClass0_);
    }

    [CompilerGenerated]
    private static void <Do>g__Help|0_0(ref <>c__DisplayClass0_0 P_0)
    {
        P_0.sb.Append("what");
    }
}
```

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs [hunk 4]
```diff
@@ -38,66 +40,357 @@ public UriPathParser(ODataUriParserSettings settings)
...
+                    startIndex = basePath.Length;
+                }
+                else
+                {
+                    // Should not happen if base-of passed and both are absolute
+                    throw new ODataException(SRResources.UriQueryPathParser_RequestUriDoesNotHaveTheCorrectBaseUri);
+                }
+
+                List<string> segments = new List<string>(4);
+                StringBuilder sb = new StringBuilder(Math.Min(256, Math.Max(0, fullPath.Length - startIndex)));
```
Comment (habbes) [inline src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs]: nit: Probably could have been a good candidate for `stackalloc char[]` or pooled array if it's large since it's short-lived. I wonder if it would be worthwhile to pool the `StringBuilder` Or to build stack-based `StringBuilder` that falls back to the heap when it grows passed a certain threshold.

Not a blocker.

File: src/Microsoft.OData.Core/UriUtils.cs [hunk 5]
```diff
@@ -229,37 +273,19 @@ internal static bool TryUriStringToTimeOfDay(ReadOnlySpan<char> text, out TimeOf
...
+            return uri.IsAbsoluteUri ? uri : new Uri(DefaultMockBase, uri);
         }
 
-        /// <summary>Creates a URI suitable for host-agnostic comparison purposes.</summary>
-        /// <param name="uri">URI to compare.</param>
-        /// <returns>URI suitable for comparison.</returns>
-        private static Uri CreateBaseComparableUri(Uri uri)
+        private static ReadOnlySpan<char> TrimSingleLeadingSlash(ReadOnlySpan<char> input)
         {
-            Debug.Assert(uri != null, "uri != null");
```
Comment (habbes) [inline src/Microsoft.OData.Core/UriUtils.cs:254]: We don't need this anymore?

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs (hunk 2 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:56]: @habbes With the previous logic, the following valid scenario would result in an unhandled exception when `UriBuilder` is invoked from https://github.com/OData/odata.net/blob/6d4b51258cc8b8d9700bd86bb13639b2f0e1d0e2/src/Microsoft.OData.Core/UriUtils.cs#L252:

```csharp
var uri = new Uri("api/People('Foo/Bar')", UriKind.RelativeOrAbsolute);
var baseUri = new Uri("api/", UriKind.RelativeOrAbsolute);

var segments = new UriPathParser().ParsePathIntoSegments(uri, baseUri);
```

This behavior was not intentional and, in my view, constitutes a bug. The change you've highlighted addresses this issue by preventing the exception, which improves robustness. As such, I wouldn’t consider this a breaking change, but rather a bug fix that aligns the behavior with expected usage.

File: src/Microsoft.OData.Core/UriUtils.cs (hunk 5 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/UriUtils.cs:254]: @habbes I got rid of the `CreateBaseComparableUri` method altogether. It was only being called from `UriInvariantInsensitiveIsBaseOf` and I refactored that logic so it doesn't depend on `UriBuilder`

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs (hunk 3 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:304]: @habbes I'm not sure how I feel about this:
```csharp
static void AddSegmentIfAny(List<string> segmentsInner, StringBuilder sbInner, int maxSegmentsInner, bool hasEncodedCharsInner)
{
    if (sbInner.Length == 0)
    {
        return;
    }

    // Depending on internal implementation, Uri.UnescapeDataString when there's no
    // escaping might result into two strings so we optimize for that case, i.e.,
    // one allocation for sbInner.ToString() and for Uri.UnescapeDataString
    string raw = sbInner.ToString();
    // Unescape percent-encoded characters for the final segment if necessary
    string segment = hasEncodedCharsInner ? Uri.UnescapeDataString(raw) : raw;

    // Skip empties and any stray "/" segment
    if (segment.Length > 0 && segment != "/")
    {
        if (segmentsInner.Count >= maxSegmentsInner)
        {
            throw new ODataException(SRResources.UriQueryPathParser_TooManySegments);
        }

        segmentsInner.Add(segment);
    }

    sbInner.Clear();
    hasEncodedCharsInner = false; // Reset for next segment
}
```

Then update two call points as follows:
```csharp
AddSegmentIfAny(segments, sb, this.maxSegments, hasEncodedChars);
```

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs [hunk 6]
```diff
@@ -38,66 +40,357 @@ public UriPathParser(ODataUriParserSettings settings)
...
+                        segments.Add(segment);
+                    }
+
+                    sb.Clear();
+                    hasEncodedChars = false; // Reset for next segment
+                }
+
+                // Walk the path after the basePath and split on unquoted '/'
+                // We don't want something like api/a'/b'/c to give us a'/b' and c instead of a', b' and c
+                for (int i = startIndex; i < fullPath.Length;)
```
Comment (habbes) [inline src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:335]: nit: I'd expect the helper local functions to be the last thing in the method definition, a bit strange to have other top-level code after the local functions. I feel like this makes the flow of the execution harder to digest.

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs (hunk 3 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:304]: Here are the result of the benchmarks I created after making AddSegmentsIfAny static and passing parameters into it:

```markdown
| Method        | Path                 | Mean      | Error    | StdDev    | Median    | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|-------------- |--------------------- |----------:|---------:|----------:|----------:|------:|--------:|-------:|----------:|------------:|
| Parse_Revised | Customers            | 107.25 ns | 5.355 ns | 14.929 ns | 103.67 ns |  1.02 |    0.19 | 0.0253 |     320 B |        1.00 |
| Parse_Latest  | Customers            |  79.35 ns | 1.624 ns |  1.519 ns |  78.97 ns |  0.75 |    0.10 | 0.0254 |     320 B |        1.00 |
|               |                      |           |          |           |           |       |         |        |           |             |
| Parse_Revised | Customers('abc')     | 105.81 ns | 2.461 ns |  7.060 ns | 102.66 ns |  1.00 |    0.09 | 0.0286 |     360 B |        1.00 |
| Parse_Latest  | Customers('abc')     | 104.49 ns | 2.900 ns |  8.504 ns | 100.53 ns |  0.99 |    0.10 | 0.0286 |     360 B |        1.00 |
|               |                      |           |          |           |           |       |         |        |           |             |
| Parse_Revised | Customers('abc/pqr') | 110.45 ns | 2.369 ns |  6.836 ns | 108.12 ns |  1.00 |    0.08 | 0.0305 |     384 B |        1.00 |
| Parse_Latest  | Customers('abc/pqr') | 110.94 ns | 1.729 ns |  3.029 ns | 110.45 ns |  1.01 |    0.06 | 0.0305 |     384 B |        1.00 |
|               |                      |           |          |           |           |       |         |        |           |             |
| Parse_Revised | Custo(...)pqr') [22] | 152.80 ns | 4.041 ns | 11.595 ns | 146.87 ns |  1.01 |    0.10 | 0.0370 |     464 B |        1.00 |
| Parse_Latest  | Custo(...)pqr') [22] | 147.81 ns | 4.002 ns | 11.675 ns | 144.40 ns |  0.97 |    0.10 | 0.0370 |     464 B |        1.00 |
|               |                      |           |          |           |           |       |         |        |           |             |
| Parse_Revised | Custo(...)Items [37] | 184.49 ns | 3.968 ns | 11.385 ns | 179.55 ns |  1.00 |    0.08 | 0.0427 |     536 B |        1.00 |
| Parse_Latest  | Custo(...)Items [37] | 180.64 ns | 3.439 ns |  8.173 ns | 177.73 ns |  0.98 |    0.07 | 0.0427 |     536 B |        1.00 |
|               |                      |           |          |           |           |       |         |        |           |             |
| Parse_Revised | Customers(1)         |  95.49 ns | 1.889 ns |  5.009 ns |  93.43 ns |  1.00 |    0.07 | 0.0267 |     336 B |        1.00 |
| Parse_Latest  | Customers(1)         |  96.51 ns | 2.684 ns |  7.527 ns |  93.34 ns |  1.01 |    0.09 | 0.0267 |     336 B |        1.00 |
|               |                      |           |          |           |           |       |         |        |           |             |
| Parse_Revised | Peopl(...)rs(1) [35] | 184.55 ns | 2.484 ns |  1.939 ns | 184.44 ns |  1.00 |    0.01 | 0.0432 |     544 B |        1.00 |
| Parse_Latest  | Peopl(...)rs(1) [35] | 192.02 ns | 3.569 ns |  3.164 ns | 191.20 ns |  1.04 |    0.02 | 0.0432 |     544 B |        1.00 |
```

Where `Parse_Latest` represents the version of `ParsePathIntoSegments` with a static `AddSegmentsIfAny`.

There's no clear winner between the two approaches and the differences can interpreted as noise given the inconsistencies.

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs (hunk 6 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs:335]: Following the investigation we did, the current placement of the local functions is the most recommended.

File: src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs (hunk 4 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs]: I’d prefer to keep the current logic as-is for now, since it prioritizes clarity and maintainability. I’d only consider this kind of macro-optimization if we had benchmark data showing a significant performance benefit.

Comment (gathogojr) [reply src/Microsoft.OData.Core/UriParser/Parsers/UriPathParser.cs]: @habbes Note also that we're using a single `StringBuilder` that we `.Clear()` after parsing a segment

**Review by habbes — APPROVED** [review]

**Review by marabooy — APPROVED** [review]
